### DISCUSSION TITLE
Modified Source and compiling doesn't have effect on device (haspmota)

### DISCUSSION THREAD

**Discussion by Silverstar** [arendst/Tasmota#18203] [Q&A]
I have a Lanbon L8 device and want to build an interface with haspmota.
I created a switch in the pages.jsonl and added val_rule:power1#state to it, so the switch follows if the relay is toggled.

As far as I know, I have to make the "switch toggle -> relay toggles" function myself, so I tried to work with tasmota.add_rule("hasp#p1b1#event==changed", ... ) but there is an issue with the event created, but this happens also without any rules defined.

In the console (weblog 4) I see an error:
`BRY: ERROR, bad json: {"hasp":{"p1b1":{"event":"changed","val":,"text":"Relay 1"}}}`
when I touch and flip the switch.
Somehow the actual value is not added to the line.

pages.jsonl:
```
{ "page": 0, "id": 0, "bg_color": "#000000", "text_color": "#FFFFFF" }

{ "page": 1, "id": 0, "bg_color": "#AAAAAA", "text_color": "#000000" }
{ "page": 1, "id": 1, "obj":"switch", "x": 37,  "y": 120, "w": 120, "h": 28, "text":"Relay 1", "val_rule":"POWER1#State", "bg_color":"#1fa3ec"}
```

I researched for the relevant source and found [haspmota.be](https://github.com/arendst/Tasmota/blob/development/tasmota/berry/haspmota_src/haspmota_core/haspmota.be#L345).
I tried to fix it myself, but no matter what I change (and I do gen_haspmota.sh afterwards!), there is no change in behaviour on the device.
(I'm working with the development branch)

I added plain output of val, commented the text:value part, but neither val is printed nor text:Relay 1 vanishes from the output, so my changes don't reflect in the compiled image.
I removed cache and .pio folders, did a clean (all) before compiling, no change, and now I am at my wit's end.

Am I working on the wrong file? What do I have to do to compile that changes in?

**Comment by sfromis** [comment]
If you want to check your Berry code change being in effect or not, you could insert print statements to document where execution got to, and maybe also add "interesting" values.

According to the [docs](https://tasmota.github.io/docs/Berry/#rules-operators), using `tasmota.add_rule("hasp#p1b1#event==changed",` is not valid, as the comparison operator `==` is for numerical comparisons. For a string comparison, use `=` instead.

When it comes to your "invalid JSON", please copy/paste the **exact** message, as what you posted looks pretty bad, what you saw in the log was probably "less bad", but different.

**Reply by Silverstar** [reply]
I tried to add prints, but it did not have any effect, as well as commenting out other lines which should have removed the "text:Relay 1" part of that message.
I will provide the exact line, but it is basically the line I posted from memory but correctly set "", but still invalid due to `val:,` with no value attached.

The same happens if I trigger on "hasp#p1b1#event", I guess it doesn't matter if I have any rule defined, the event happens regardless.

Thank you for coming by nonetheless, I will update with the exact log line and jsonl.

**Comment by s-hadinger** [comment]
Can you please paste the exact error from the console. It’s hard to analyze json behavior if we don’t have the exact line

**Reply by s-hadinger** [reply]
For reasons due to lvgl symbols, the haspmota.be solidification needs to be done on an actual Tasmota device using the lines commented at the beginning of the file. It's  it super easy

**Reply by Silverstar** [reply]
@sfromis, @s-hadinger, I updated the first post with the exact log line and my simple test pages.jsonl. 
This occurs, on a bare tasmota32-lvgl development v12.4.0.2, after a device reset, no rules defined, restarted. No matter if precompiled or my own.

**Reply by Silverstar** [reply]
> For reasons due to lvgl symbols, the haspmota.be solidification needs to be done on an actual Tasmota device using the lines commented at the beginning of the file. It's it super easy

So even flashing a new image it has to be _manually_ solidified otherwise the old .bec stays somewhere in the flash?

**Reply by s-hadinger** [reply]
Yes indeed. You can however force a local version by donwloading `haspmota.be` and doing `load("haspmota.be")` it will supersede the version in Flash

**Reply by Silverstar** [reply]
Except solidifying manually, shouldn't a `reset 1/2/3/...` remove all such solidified files? Especially when "erase all flash" is stated.
And flashing a new image with a new source file should issue a new solidification, imho.
Because I would expect flashing a new system image is overwriting system files, and I count haspmota.be as such, and the .bec to not be included in the image, so it will be solidified on first boot.

Well, nevertheless, on the bug topic, the event is broken; I don't see the reason at first glance when looking at the source, do you?
I would have done some try and error if I got that solidify stuff right.

**Reply by sfromis** [reply]
No, solidification is not the `.bec` file, but generation of code to be included when building the Tasmota binary.

To test with your version, the `load` suggested by s-hadinger is probably to most practical, needs no compile of a Tasmota binary at all.

**Reply by Silverstar** [reply]
I assumed it would be haspmota.bec because the solidification example in the header removes it before solidification. Sorry, I'm unfamiliar with solidification.
If I find some time next week I'll try that.
Thank you for all that elaboration.

I'm not sure if it's really within the haspmota.be, because I found #18133 in which the switch get_val got a changed parameter -1 instead of 1, which I don't understand yet: 
[line -4141 +4384](https://github.com/arendst/Tasmota/commit/7167884f0635c9e353979d1eaefe129c76b9b452#diff-25981299f7293bab200ff5d79a949e729e149beb45d3c7cea4259e9024a89c4bR4384)

**Reply by sfromis** [reply]
If you want to check for differences, you should not look at the solidified code which is what you are linking to. Hard to read, unless you really dive into the details of how solidification works, and doubt that the solidification works properly. In this specific case, do not look at `be_lv_haspmota.c` but the original Berry code in `haspmota.be`.
